### PR TITLE
More livequery fixes

### PR DESCRIPTION
**Pull request by millerdev**
Finding more holes in livequery and patching them. Best reviewed by commit 🐠 

@ctsims, will be submitting new test cases in this PR to mobile.

@proteusvacuum @snopoke @dannyroberts

### CODE REVIEW

**Comment by snopoke** [comment]
Will try find some headspace tomorrow to review this.

**Comment by millerdev** [comment]
Working on test failure.
```
FAIL: casexml.apps.phone.tests.test_sync_mode:LiveQuerySyncTokenUpdateTestSQL.test_cousins
----------------------------------------------------------------------
Traceback (most recent call last):
  File "corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py", line 775, in test_cousins
    child_2.case_id
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 114, in assert_user_has_cases
    should_have=True, line_by_line=False, **kwargs)
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 166, in check_user_has_case
    with cached_restore(testcase, user, restore_id, version, purge_restore_cache):
  File "/usr/local/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 151, in cached_restore
    testcase.payload_string = testcase.restore_config.get_payload().as_string()
  File "corehq/ex-submodules/casexml/apps/phone/restore.py", line 703, in get_payload
    response = self.generate_payload()
  File "corehq/ex-submodules/casexml/apps/phone/restore.py", line 709, in generate_payload
    response = self._generate_restore_response(async_task=async_task)
  File "corehq/ex-submodules/casexml/apps/phone/restore.py", line 771, in _generate_restore_response
    partial_response = provider.get_response(self.restore_state)
  File "corehq/ex-submodules/casexml/apps/phone/data_providers/case/providers.py", line 13, in get_response
    return do_livequery(self.timing_context, restore_state, self.async_task)
  File "corehq/ex-submodules/casexml/apps/phone/data_providers/case/livequery.py", line 204, in do_livequery
    for index in related
  File "corehq/ex-submodules/casexml/apps/phone/data_providers/case/livequery.py", line 206, in <setcomp>
    and index.case_id not in deleted_ids}
  File "corehq/ex-submodules/casexml/apps/phone/data_providers/case/livequery.py", line 115, in classify
    assert ix_key not in seen_ix[sub_id], ix_key
AssertionError: Stannis parent
-------------------- >> begin captured logging << --------------------
corehq.form_processor.tests.utils: DEBUG: Deleting all Couch cases for domain None
corehq.form_processor.tests.utils: DEBUG: Deleting all SQL cases for domain None
corehq.form_processor.tests.utils: DEBUG: Deleting all Couch xforms for domain None
corehq.form_processor.tests.utils: DEBUG: Deleting all SQL xforms for domain None
corehq.form_processor.tests.utils: DEBUG: Deleting all synclogs
casexml.apps.phone.data_providers.case.livequery: DEBUG: sync 5089a44138c383b8973d3d5aed909d40 for [u'3043a0086aff48b2a8e94ab198a89c30']
casexml.apps.phone.data_providers.case.livequery: DEBUG: owned: []
casexml.apps.phone.data_providers.case.livequery: DEBUG: open: set([])
casexml.apps.phone.data_providers.case.livequery: DEBUG: live: set([])
casexml.apps.phone.models: DEBUG: updating sync log for 3043a0086aff48b2a8e94ab198a89c30
casexml.apps.phone.models: DEBUG: case ids before update: 
casexml.apps.phone.models: DEBUG: dependent case ids before update: 
casexml.apps.phone.models: DEBUG: index tree before update: {}
casexml.apps.phone.models: DEBUG: extension index tree before update: {}
casexml.apps.phone.models: DEBUG: initializing update for case Stannis
casexml.apps.phone.models: DEBUG: Stannis: create
casexml.apps.phone.models: DEBUG: Stannis: update
casexml.apps.phone.models: DEBUG: Stannis: index
casexml.apps.phone.models: DEBUG: Stannis: create
casexml.apps.phone.models: DEBUG: Stannis: update
casexml.apps.phone.models: DEBUG: Stannis: index
casexml.apps.phone.models: DEBUG: initializing update for case Robert
casexml.apps.phone.models: DEBUG: Robert: create
casexml.apps.phone.models: DEBUG: Robert: update
casexml.apps.phone.models: DEBUG: Robert: index
casexml.apps.phone.models: DEBUG: Robert: create
casexml.apps.phone.models: DEBUG: Robert: update
casexml.apps.phone.models: DEBUG: Robert: index
casexml.apps.phone.models: DEBUG: initializing update for case Shireen
casexml.apps.phone.models: DEBUG: Shireen: create
casexml.apps.phone.models: DEBUG: Shireen: update
casexml.apps.phone.models: DEBUG: Shireen: index
casexml.apps.phone.models: DEBUG: initializing update for case Joffrey
casexml.apps.phone.models: DEBUG: Joffrey: create
casexml.apps.phone.models: DEBUG: Joffrey: update
casexml.apps.phone.models: DEBUG: Joffrey: index
casexml.apps.phone.models: DEBUG: initializing update for case Steffon
casexml.apps.phone.models: DEBUG: Steffon: create
casexml.apps.phone.models: DEBUG: Steffon: update
casexml.apps.phone.models: DEBUG: Steffon: create
casexml.apps.phone.models: DEBUG: Steffon: update
casexml.apps.phone.models: DEBUG: Steffon: create
casexml.apps.phone.models: DEBUG: Steffon: update
casexml.apps.phone.models: DEBUG: Steffon: create
casexml.apps.phone.models: DEBUG: Steffon: update
casexml.apps.phone.models: DEBUG: Steffon: create
casexml.apps.phone.models: DEBUG: Steffon: update
casexml.apps.phone.models: DEBUG: case Shireen is live.
casexml.apps.phone.models: DEBUG: adding index Shireen --<child>--> Stannis (parent).
casexml.apps.phone.models: DEBUG: case Joffrey is live.
casexml.apps.phone.models: DEBUG: adding index Joffrey --<child>--> Robert (parent).
casexml.apps.phone.models: DEBUG: case Stannis is NOT live.
casexml.apps.phone.models: DEBUG: case Robert is NOT live.
casexml.apps.phone.models: DEBUG: case Steffon is NOT live.
casexml.apps.phone.models: DEBUG: case ids mid update: Stannis, Robert, Shireen, Joffrey
casexml.apps.phone.models: DEBUG: dependent case ids mid update: Stannis, Robert
casexml.apps.phone.models: DEBUG: index tree mid update: {
  "Shireen": {
    "parent": "Stannis"
  }, 
  "Joffrey": {
    "parent": "Robert"
  }
}
casexml.apps.phone.models: DEBUG: extension index tree mid update: {}
casexml.apps.phone.models: DEBUG: purging: Stannis
casexml.apps.phone.models: DEBUG: Relevant cases of Stannis: set([u'Stannis', u'Shireen'])
casexml.apps.phone.models: DEBUG: Available cases: set([u'Stannis', u'Shireen'])
casexml.apps.phone.models: DEBUG: live cases: set([u'Stannis', u'Shireen'])
casexml.apps.phone.models: DEBUG: cases to to_remove: set([])
casexml.apps.phone.models: DEBUG: adding index Stannis --<child>--> Steffon (parent).
casexml.apps.phone.models: DEBUG: adding index Stannis --<child>--> Steffon (parent).
casexml.apps.phone.models: DEBUG: purging: Robert
casexml.apps.phone.models: DEBUG: Relevant cases of Robert: set([u'Robert', u'Joffrey'])
casexml.apps.phone.models: DEBUG: Available cases: set([u'Robert', u'Joffrey'])
casexml.apps.phone.models: DEBUG: live cases: set([u'Robert', u'Joffrey'])
casexml.apps.phone.models: DEBUG: cases to to_remove: set([])
casexml.apps.phone.models: DEBUG: adding index Robert --<child>--> Steffon (parent).
casexml.apps.phone.models: DEBUG: adding index Robert --<child>--> Steffon (parent).
casexml.apps.phone.models: DEBUG: purging: Steffon
casexml.apps.phone.models: DEBUG: Relevant cases of Steffon: set([u'Steffon'])
casexml.apps.phone.models: DEBUG: Available cases: set([u'Steffon'])
casexml.apps.phone.models: DEBUG: live cases: set([])
casexml.apps.phone.models: DEBUG: cases to to_remove: set([u'Steffon'])
casexml.apps.phone.models: DEBUG: removing: Steffon
casexml.apps.phone.models: DEBUG: case ids after update: Stannis, Robert, Shireen, Joffrey
casexml.apps.phone.models: DEBUG: dependent case ids after update: Stannis, Robert
casexml.apps.phone.models: DEBUG: index tree after update: {
  "Stannis": {
    "parent": "Steffon"
  }, 
  "Robert": {
    "parent": "Steffon"
  }, 
  "Shireen": {
    "parent": "Stannis"
  }, 
  "Joffrey": {
    "parent": "Robert"
  }
}
casexml.apps.phone.models: DEBUG: extension index tree after update: {}
casexml.apps.phone.models: DEBUG: made changes, saving.
root: DEBUG: Beginning atomic commit
root: DEBUG: Saving new form: XFormInstance(form_id='f85964fbbda64fdf927e66dcdef2cd54', domain='test-domain')
casexml.apps.phone.data_providers.case.livequery: DEBUG: sync 5089a44138c383b8973d3d5aed909a1e for [u'3043a0086aff48b2a8e94ab198a89c30']
casexml.apps.phone.data_providers.case.livequery: DEBUG: owned: [u'Shireen', u'Joffrey']
casexml.apps.phone.data_providers.case.livequery: DEBUG: Shireen --child--> Stannis
casexml.apps.phone.data_providers.case.livequery: DEBUG: enliven(Shireen)
casexml.apps.phone.data_providers.case.livequery: DEBUG: enliven(Stannis)
casexml.apps.phone.data_providers.case.livequery: DEBUG: Joffrey --child--> Robert
casexml.apps.phone.data_providers.case.livequery: DEBUG: enliven(Joffrey)
casexml.apps.phone.data_providers.case.livequery: DEBUG: enliven(Robert)
casexml.apps.phone.data_providers.case.livequery: DEBUG: next: set([u'Stannis', u'Robert'])
casexml.apps.phone.data_providers.case.livequery: DEBUG: Stannis --child--> Steffon
casexml.apps.phone.data_providers.case.livequery: DEBUG: enliven(Steffon)
--------------------- >> end captured logging << ---------------------
```

File: corehq/tests/noseplugins/output.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+"""Print collected test output for passing tests
+
+Usage: ./manage.py test --with-output
```
Comment (dannyroberts) [inline corehq/tests/noseplugins/output.py]: 👍 happy about this

**Review by dannyroberts — APPROVED** [review]
Looks reasonable to me but also hard to follow changes this technical without being really deep into it. Any tips/suggestions for how to approach review? I'm also curious how much we feel like we can rely on current tests to tell us whether this is reaching functional parity with traditional sync and/or how we plan on knowing that before doing the flip. (If I had been paying closer attention in the last weeks, I might already know the answer to this question.

**Comment by ctsims** [comment]
@dannyroberts I can't speak for Daniel but I think the review here would be more for the review of the structure and quality of the code, more than for "correctness", where we're trying to rely more on expanded testing and careful rollout (since we're still behind flags).

**Comment by millerdev** [comment]
What @ctsims said. From what I've been hearing from others, we think our tests are pretty good for this code. I've put a lot of thought into edge cases and added tests for all of the scenarios I could think of that would cause livequery to generate the wrong result.

In terms of assessing correctness, I tried to add comments to explain how each branch of code is affecting the final result. I'd also be happy to do a code walk through with anyone who thinks that would be helpful.

**Comment by millerdev** [comment]
^ @dannyroberts

File: corehq/ex-submodules/casexml/apps/phone/data_providers/case/livequery.py [hunk 2]
```diff
@@ -45,8 +45,25 @@ def do_livequery(timing_context, restore_state, async_task=None):
...
         return '{} {}'.format(index.case_id, index.identifier)
 
+    def is_extension(case_id):
+        """Determine if case_id is an extension case
+
+        A case that is both a child and an extension is not an extension.
+        """
+        return case_id in hosts_by_extension and case_id not in parents_by_child
+
     def has_live_extension(case_id, cache={}):
```
Comment (millerdev) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/case/livequery.py]: yes, that's [exactly what I meant to do](http://python-guide-pt-br.readthedocs.io/en/latest/writing/gotchas/#when-the-gotcha-isn-t-a-gotcha). Python [evaluates default function parameters at the time when the function is defined](http://python-guide-pt-br.readthedocs.io/en/latest/writing/gotchas/#mutable-default-arguments).
